Makefile

```diff
@@ -269,54 +277,47 @@ images: k8s-broker-image user-broker-image \
 k8s-broker-image: contrib/build/k8s-broker/Dockerfile $(BINDIR)/k8s-broker
 	mkdir -p contrib/build/k8s-broker/tmp
 	cp $(BINDIR)/k8s-broker contrib/build/k8s-broker/tmp
-	docker build -t k8s-broker:$(VERSION) contrib/build/k8s-broker
+	docker build -t $(K8S_BROKER_IMAGE) contrib/build/k8s-broker
+	docker tag $(K8S_BROKER_IMAGE) $(K8S_BROKER_MUTABLE_IMAGE)
 	rm -rf contrib/build/k8s-broker/tmp
 
 user-broker-image: contrib/build/user-broker/Dockerfile $(BINDIR)/user-broker
 	mkdir -p contrib/build/user-broker/tmp
 	cp $(BINDIR)/user-broker contrib/build/user-broker/tmp
-	docker build -t user-broker:$(VERSION) contrib/build/user-broker
-	docker tag user-broker:$(VERSION) user-broker:latest
+	docker build -t $(USER_BROKER_IMAGE) contrib/build/user-broker
+	docker tag $(USER_BROKER_IMAGE) $(USER_BROKER_MUTABLE_IMAGE)
 	rm -rf contrib/build/user-broker/tmp
 
 apiserver-image: build/apiserver/Dockerfile $(BINDIR)/apiserver
 	mkdir -p build/apiserver/tmp
 	cp $(BINDIR)/apiserver build/apiserver/tmp
-	docker build -t apiserver:$(VERSION) build/apiserver
-	docker tag apiserver:$(VERSION) apiserver:latest
+	docker build -t $(APISERVER_IMAGE) build/apiserver
+	docker tag $(APISERVER_IMAGE) $(APISERVER_MUTABLE_IMAGE)
 	rm -rf build/apiserver/tmp
 
 controller-manager-image: build/controller-manager/Dockerfile $(BINDIR)/controller-manager
 	mkdir -p build/controller-manager/tmp
 	cp $(BINDIR)/controller-manager build/controller-manager/tmp
-	docker build -t controller-manager:$(VERSION) build/controller-manager
-	docker tag controller-manager:$(VERSION) controller-manager:latest
+	docker build -t $(CONTROLLER_MANAGER_IMAGE) build/controller-manager
+	docker tag $(CONTROLLER_MANAGER_IMAGE) $(CONTROLLER_MANAGER_MUTABLE_IMAGE)
 	rm -rf build/controller-manager/tmp
 
 # Push our Docker Images to a registry
 ######################################
 push: k8s-broker-push user-broker-push controller-manager-push apiserver-push
 
 k8s-broker-push: k8s-broker-image
-	docker tag k8s-broker:$(VERSION) $(REGISTRY)k8s-broker:$(VERSION)
-	docker push $(REGISTRY)k8s-broker:$(VERSION)
-	docker tag k8s-broker:$(VERSION) $(REGISTRY)k8s-broker:canary
-	docker push $(REGISTRY)k8s-broker:canary
+	docker push $(K8S_BROKER_IMAGE)
+	docker push $(K8S_BROKER_MUTABLE_IMAGE)
```

Comment:
@duglin each `x-push` target has `x-image` as a dependency and all the `x-image` targets are phonies. This means that upon `make push`, all images are always rebuilt anyway-- an operation which is fast because of all the caching Docker does. Long story short... regardless of when you set `REGISTRY` (before `make build` or before `make push`) you're going to end up with consistent results.